### PR TITLE
Adds a JSON web api for looking up/creating clusters.

### PR DESCRIPTION
**Pull request by xa4a**
This adds lookup methods for individual clusters and shreds by id, looking up random cluster, optionally by shard and merging clusters. With tests.

### CODE REVIEW

File: webapi.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+          },
+          ...
+        ]
+    } }
+
+    """
+
+    req = request.get_json().get('cluster')
+
+    parents = map(lambda pk: Cluster.objects.get_or_404(pk=pk), req['parents'])
```
Comment (dchaplinsky) [inline webapi.py]: Will it work without lambda? :)

Comment (xa4a) [inline webapi.py]: No :(
InvalidQueryError: Not a query object: fixtures1:6_3. Did you intend to use key=value?

File: webapi.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    cluster = Cluster(
+            id=str(uuid.uuid1()),
+            users_count=0,
+            users_skipped=[],
+            users_processed=[],
+            batch=parents[0].batch,
+            tags=[],
+            parents=parents,
+            members=[ClusterMember(
+                shred=m['shred'], position=m['position'], angle=m['angle'])
```
Comment (dchaplinsky) [inline webapi.py]: Shouldn't we check for the presence of this params before diving into objects creation?

File: webapi.py (hunk 1 above)
Comment (xa4a) [inline webapi.py]: Well, can be list comprehension.

File: app.py [hunk 3]
```diff
@@ -6,12 +6,14 @@
 
 from flask.ext.mongoengine import MongoEngine
 from flask.ext import login
+from werkzeug.wsgi import DispatcherMiddleware
```
Comment (dchaplinsky) [inline app.py]: Not used?

File: webapi.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        A JSON response like:
+        { "success": True,
+          "data": [Cluseter1(), ..., ClusterN()],
+        }
+        If the cluster for the requested id is missing, null is returned in
+        corresponding position.
+    """
+    ids = list(request.get_json().get('ids'))
+
+    mapping = dict((c.id, c) for c in Cluster.objects.filter(id__in=ids))
```
Comment (dchaplinsky) [inline webapi.py]: Why not json_mapping right from the start?

Comment (dchaplinsky) [inline webapi.py]: What if ids contains wrong id?

File: tests/webapi_test.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        resp = self.client.get(
+            url_for('webapi.get_cluster')).json['data']
+        cluster1_id = resp['cluster']['_id']
+        cluster1_batch = resp['cluster']['batch']
+        cluster2_id = cluster1_id
+
+        qs = urllib.urlencode({'batch': cluster1_batch})
+
+        panic_counter = 0
+        while cluster2_id == cluster1_id:
```
Comment (dchaplinsky) [inline tests/webapi_test.py]: Awfuck. ugly, ugly!

File: tests/webapi_test.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            # Sometimes it just won't give another cluster.
+            if panic_counter > 100:
+                return self._get_mergeable_clusters()
```
Comment (dchaplinsky) [inline tests/webapi_test.py]: Non-zero chance of stack overflow!

File: webapi.py (hunk 4 above)
Comment (dchaplinsky) [inline webapi.py]: Hmm, ok

File: webapi.py (hunk 2 above)
Comment (xa4a) [inline webapi.py]: Done.

**Pull request merged**